sonar-home/src/main/java/org/sonar/home/cache/DirectoryLock.java

```diff
@@ -0,0 +1,106 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.home.cache;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.RandomAccessFile;
+import java.io.StringWriter;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class DirectoryLock {
+  static final String LOCK_FILE_NAME = ".sonar_lock";
+  private final Path lockFilePath;
+  private final Logger logger;
+
+  private RandomAccessFile lockRandomAccessFile;
+  private FileChannel lockChannel;
+  private FileLock lockFile;
+
+  public DirectoryLock(Path directory, Logger logger) {
+    this.logger = logger;
+    this.lockFilePath = directory.resolve(LOCK_FILE_NAME).toAbsolutePath();
+  }
+
+  public String getFileLockName() {
+    return LOCK_FILE_NAME;
+  }
+  
+  public void lock() {
+    try {
+      lockRandomAccessFile = new RandomAccessFile(lockFilePath.toFile(), "rw");
+      lockChannel = lockRandomAccessFile.getChannel();
+      lockFile = lockChannel.lock(0, 1024, false);
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to create lock in " + lockFilePath.toString(), e);
+    }
+  }
+  
+  public boolean tryLock() {
+    try {
+      lockRandomAccessFile = new RandomAccessFile(lockFilePath.toFile(), "rw");
+      lockChannel = lockRandomAccessFile.getChannel();
+      lockFile = lockChannel.tryLock(0, 1024, false);
+
+      return lockFile != null;
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to create lock in " + lockFilePath.toString(), e);
+    }
+  }
+
+  public void unlock() {
+    if (lockFile != null) {
+      try {
+        lockFile.release();
+        lockFile = null;
+      } catch (IOException e) {
+        logger.error("Error releasing lock", e);
+      }
+    }
+    if (lockChannel != null) {
+      try {
+        lockChannel.close();
+        lockChannel = null;
+      } catch (IOException e) {
+        logger.error("Error closing file channel", e);
+      }
+    }
+    if (lockRandomAccessFile != null) {
+      try {
+        lockRandomAccessFile.close();
+        lockRandomAccessFile = null;
+      } catch (IOException e) {
+        logger.error("Error closing file", e);
+      }
+    }
+
+    try {
+      Files.delete(lockFilePath);
+    } catch (IOException e) {
+      // ignore, as an error happens if another process just started to acquire the same lock
+      StringWriter errors = new StringWriter();
+      e.printStackTrace(new PrintWriter(errors));
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1148)